src/vs/vscode.d.ts

```diff
@@ -8940,6 +8940,267 @@ declare module 'vscode' {
+	}
+
+	export interface CommentThreadTemplate {
+		/**
+		 * The human-readable label describing the [Comment Thread](#CommentThread)
+		 */
+		label: string;
+
+		/**
+		 * Optional accept input command
+		 *
+		 * `acceptInputCommand` is the default action rendered on Comment Widget, which is always placed rightmost.
+		 * This command will be invoked when users the user accepts the value in the comment editor.
+		 * This command will disabled when the comment editor is empty.
+		 */
+		acceptInputCommand?: Command;
+
+		/**
+		 * Optional additonal commands.
+		 *
+		 * `additionalCommands` are the secondary actions rendered on Comment Widget.
+		 */
+		additionalCommands?: Command[];
+
+		/**
+		 * The command to be executed when users try to delete the comment thread. Currently, this is only called
+		 * when the user collapses a comment thread that has no comments in it.
+		 */
+		deleteCommand?: Command;
+	}
+
+	export interface EmptyCommentThreadFactory {
+		template: CommentThreadTemplate;
+		/**
+		 * When users attempt to create new comment thread from the gutter or command palette, `template` will be used first to create the Comment Thread Widget in the editor for users to start comment drafting.
+		 * Then `createEmptyCommentThread` is called after that. Extensions should still call [`createCommentThread`](CommentController.createCommentThread) to create a real [`CommentThread`](#CommentThread)
+		 * Extensions still need to call `createCommentThread` inside this call when appropriate.
+		 *
+		 * @param document The document in which users attempt to create a new comment thread
+		 * @param range The range the comment threadill located within the document.
+		 *
+		 * @returns commentThread The [`CommentThread`](#CommentThread) created by extensions
+		 */
+		createEmptyCommentThread(document: TextDocument, range: Range): ProviderResult<CommentThread>;
+	}
+
+	/**
+	 * A comment controller is able to provide [comments](#CommentThread) support to the editor and
+	 * provide users various ways to interact with comments.
+	 */
+	export interface CommentController {
+		/**
+		 * The id of this comment controller.
+		 */
+		readonly id: string;
+
+		/**
+		 * The human-readable label of this comment controller.
+		 */
+		readonly label: string;
+
+		/**
+		 * The active [comment input box](#CommentInputBox) or `undefined`. The active `inputBox` is the input box of
+		 * the comment thread widget that currently has focus. It's `undefined` when the focus is not in any CommentInputBox.
+		 */
+		readonly inputBox: CommentInputBox | undefined;
+
+		/**
+		 * The active [comment thread](#CommentThread) or `undefined`. The `activeCommentThread` is the comment thread of
+		 * the comment thread widget that currently has focus. It's `undefined` when the focus is not in any comment thread widget.
+		 */
+		readonly activeCommentThread: CommentThread | undefined;
```

Comment:
Having this but no `CommentController#commentThreads: CommentThread[]` is slightly weird and seems to mix to approaches, e.g one approach is more push-like with extensions calling `createCommentThread` while the other would be more pull-like where the controller provides "all" comment threads (which the UX picks up and denotes one as the active one)

Reply:
I'm wondering if having both `CommentController#activeCommentThread` and `CommentController#commentThreads` might be confusing as `activeCommentThread` is set by VS Code but `commentThreads` are set by extensions.